minigui/gtp_socket.ts

```diff
@@ -0,0 +1,122 @@
+// Let's not worry about Socket.io type safety for now.
+declare var io: any;
+
+type DebugHandler = (msg: string) => void;
+type CmdHandler = (cmd: string, result: string) => void;
+type StderrHandler = (line: string) => void;
+type ConnectCallback = () => void;
+
+// The GtpSocket serializes all calls to send so that only one call is ever
+// outstanding at a time. This isn't strictly necessary, but it makes reading
+// the debug logs easier because we don't end up with request and result logs
+// all jumbled up.
+class Socket {
+  private sock: any;
+  private cmdQueue = new Array<{cmd: string, callback: CmdHandler | null}>();
+  private gameToken: string;
+  private handshakeComplete = false;
+  private connectCallback: ConnectCallback | null = null;
+
+  private stderrHandlers = new Array<{prefix: string, handler: StderrHandler}>();
+
+  constructor(uri: string, private debugHandler: DebugHandler) {
+    this.sock = io.connect(uri)
+    this.debugHandler = debugHandler;
+
+    this.sock.on('connect', () => {
+      this.newSession();
+      if (this.connectCallback) {
+        this.connectCallback();
+      }
+     });
+
+    this.sock.on('json', (msg: string) => {
+      let obj = JSON.parse(msg);
+      if (obj.token != this.gameToken) {
+        console.log('ignoring', obj, `${obj.token} != ${this.gameToken}`);
+        return;
+      }
+
+      if (obj.stdout !== undefined) {
+        if (obj.stdout.trim() != '') {
+          this.cmdHandler(obj.stdout);
+        }
+      } else if (obj.stderr !== undefined) {
+        this.stderrHandler(obj.stderr);
+      }
+    });
+  }
+
+  onConnect(callback: ConnectCallback) {
+    this.connectCallback = callback;
+  }
+
+  addStderrHandler(prefix: string, handler: StderrHandler) {
+    this.stderrHandlers.push({prefix: prefix, handler: handler});
+  }
+
+  send(cmd: string, callback?: CmdHandler) {
+    this.cmdQueue.push({cmd: cmd, callback: callback || null});
+    if (this.cmdQueue.length == 1) {
+      this.sendNext();
+    }
+  }
+
+  newSession() {
+    this.cmdQueue = [];
+    let token = `session-id-${Date.now()}`;
+    this.gameToken = token;
+    let cmd = `echo __NEW_TOKEN__ ${token}`;
+    this.sock.emit('gtpcmd', {data: cmd});
+  }
+
+  private cmdHandler(result: string) {
+    let {cmd, callback} = this.cmdQueue[0];
+    this.cmdQueue = this.cmdQueue.slice(1);
+
+    if (this.debugHandler) {
+      let trimmedCmd = cmd;
+      if (trimmedCmd.substr(0, 3) == 'mg_') {
+        trimmedCmd = cmd.substr(3);
```

Comment:
i think the behavior here might end up with folks thinking those are the GTP commands they need to type; i.e., someone who knows a bit about GTP will type "genmove b" and then wonder why no variations appear.  I totally didn't do this myself a few times.... of course not....

Reply:
done